# <font color="red">**3. L1 FS - Fine-tuning Hyperparameters**</font>

**Author:** Osmar Bolivar

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
import plotly.graph_objects as go

## **1. Monthly data**

In [2]:
raw = pd.read_csv("./Data/RF_DATASET.csv", index_col=0)
raw.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5901 entries, 2010-01-01 to 2024-10-31
Data columns (total 54 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   ipc_all       178 non-null    float64
 1   lag_1         5901 non-null   float64
 2   lag_2         5901 non-null   float64
 3   lag_3         5901 non-null   float64
 4   lag_6         5901 non-null   float64
 5   lag_9         5901 non-null   float64
 6   lag_12        5901 non-null   float64
 7   oil_cb        5901 non-null   float64
 8   sugar_bol     5901 non-null   float64
 9   chicken_tj    5901 non-null   float64
 10  oil2_tj       5901 non-null   float64
 11  oil2_bol      5901 non-null   float64
 12  sugar_tj      5901 non-null   float64
 13  rice4_tj      5901 non-null   float64
 14  sugar_cb      5901 non-null   float64
 15  rice3_po      5901 non-null   float64
 16  sugar_or      5901 non-null   float64
 17  sugar_sc      5901 non-null   float64
 18  rice4_su      5901

In [3]:
df = raw.copy().query('freq == "month"')
df.drop(columns=["freq"], inplace=True)
df

ipc_all     lag_1     lag_2     lag_3     lag_6     lag_9  \
2010-01-31  0.074472  0.263790  0.457498  0.784931  1.447299  5.323961   
2010-02-28  0.312381  0.074472  0.263790  0.457498  1.417620  3.200128   
2010-03-31  0.685510  0.312381  0.074472  0.263790  0.643999  2.117642   
2010-04-30  1.215574  0.685510  0.312381  0.074472  0.784931  1.447299   
2010-05-31  1.384706  1.215574  0.685510  0.312381  0.457498  1.417620   
...              ...       ...       ...       ...       ...       ...   
2024-06-30  3.844356  3.516994  3.460018  3.063085  2.118325  2.847826   
2024-07-31  3.953707  3.844356  3.516994  3.460018  1.863150  2.075432   
2024-08-31  5.192869  3.953707  3.844356  3.516994  2.516604  1.594145   
2024-09-30  6.182067  5.192869  3.953707  3.844356  3.063085  2.118325   
2024-10-31  7.938964  6.182067  5.192869  3.953707  3.460018  1.863150   

               lag_12     oil_cb   sugar_bol  chicken_tj  ...  chicken_sc  \
2010-01-31  11.041182   8.829032  173.290323   12.822581  ...    9.494194   
2010-02-28   8.128586   8.800000  183.071429   12.714286  ...   11.184286   
2010-03-31   6.563669   8.800000  180.000000   13.112903  ...   11.458065   
2010-04-30   5.323961   8.800000  184.833333   13.500000  ...   12.020000   
2010-05-31   3.200128   8.800000  196.451613   13.032258  ...   10.316129   
...               ...        ...         ...         ...  ...         ...   
2024-06-30   2.725282  10.500000  245.000000   13.500000  ...   13.641667   
2024-07-31   2.700872  10.478710  242.161290   13.935484  ...   13.751613   
2024-08-31   3.058496  11.132258  241.161290   14.000000  ...   13.932258   
2024-09-30   2.847826  11.660000  242.833333   14.000000  ...   13.316667   
2024-10-31   2.075432  11.701290  242.580645   14.387097  ...   15.012258   

            tomato_co     libor  platano_co    sugar_po  carrot_co  tomato_tj  \
2010-01-31      174.0  0.400828   75.000000  174.032258  62.500000  44.303871   
2010-02-28      174.0  0.387948   75.000000  184.107143  62.500000  53.852143   
2010-03-31      174.0  0.411916   75.000000  183.322581  62.500000  46.201290   
2010-04-30      174.0  0.470366   75.666667  188.300000  62.500000  44.731000   
2010-05-31      174.0  0.655943   71.612903  196.290323  58.870968  38.882258   
...               ...       ...         ...         ...        ...        ...   
2024-06-30      110.0  5.709759   66.670000  240.000000  55.000000  94.752000   
2024-07-31      110.0  5.590282   66.670000  240.000000  55.000000  81.698387   
2024-08-31      110.0  5.238481   66.670000  240.000000  55.000000  94.300000   
2024-09-30      110.0  4.939704   66.670000  240.000000  55.000000  94.300000   
2024-10-31      110.0  4.682387   66.670000  240.000000  55.000000  92.127419   

               oil_tr  lard_lp  rice4_or  
2010-01-31   9.170000    185.0    250.00  
2010-02-28   9.170000    185.0    250.00  
2010-03-31   9.170000    185.0    250.00  
2010-04-30   9.169333    185.0    250.00  
2010-05-31   8.931290    185.0    250.00  
...               ...      ...       ...  
2024-06-30  10.800000    232.5     10.17  
2024-07-31  10.800000    232.5     10.17  
2024-08-31  10.800000    232.5     10.17  
2024-09-30  10.800000    232.5     10.17  
2024-10-31  10.800000    232.5     10.17  

[178 rows x 53 columns]

In [4]:
isna = df.isna().sum()
isna[isna > 0]

Series([], dtype: int64)

In [5]:
df.shape

(178, 53)

## **2. Train and Val sets**

In [6]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [7]:
train_size = int(len(df) * 0.8)
train, validation = df[:train_size], df[train_size:]
print(f'Obs in train set: {train.shape[0]}; variables in train set: {train.shape[1]}')
print(f'Obs in validation set: {validation.shape[0]}; variables in validation set: {validation.shape[1]}')

Obs in train set: 142; variables in train set: 53
Obs in validation set: 36; variables in validation set: 53


In [8]:
validation.info()

<class 'pandas.core.frame.DataFrame'>
Index: 36 entries, 2021-11-30 to 2024-10-31
Data columns (total 53 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   ipc_all       36 non-null     float64
 1   lag_1         36 non-null     float64
 2   lag_2         36 non-null     float64
 3   lag_3         36 non-null     float64
 4   lag_6         36 non-null     float64
 5   lag_9         36 non-null     float64
 6   lag_12        36 non-null     float64
 7   oil_cb        36 non-null     float64
 8   sugar_bol     36 non-null     float64
 9   chicken_tj    36 non-null     float64
 10  oil2_tj       36 non-null     float64
 11  oil2_bol      36 non-null     float64
 12  sugar_tj      36 non-null     float64
 13  rice4_tj      36 non-null     float64
 14  sugar_cb      36 non-null     float64
 15  rice3_po      36 non-null     float64
 16  sugar_or      36 non-null     float64
 17  sugar_sc      36 non-null     float64
 18  rice4_su      36 non

In [9]:
scaler = StandardScaler()
train_scaled = pd.DataFrame(scaler.fit_transform(train), columns=train.columns, index=train.index)
validation_scaled = pd.DataFrame(scaler.transform(validation), columns=validation.columns, index=validation.index)
#test_scaled = pd.DataFrame(scaler.transform(test), columns=test.columns, index=test.index)

X_train = train_scaled.drop('ipc_all', axis=1)
y_train = train_scaled['ipc_all']

X_validation = validation_scaled.drop('ipc_all', axis=1)
y_validation = validation_scaled['ipc_all']

#X_test = test_scaled.drop('ipc_all', axis=1)

## **3. Algorithms**

In [10]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import TimeSeriesSplit
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso
from sklearn.linear_model import ElasticNet
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import AdaBoostRegressor
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import ExtraTreesRegressor

tscv5 = TimeSeriesSplit(n_splits=5)

### **3.1. Ridge**   

**Without tuning (Validation):**
MSE:  0.178; R2:  0.436; MAE:  0.273

In [52]:
#ridge = Ridge(random_state=0)
ridge = Ridge(alpha=3.4, fit_intercept=False, positive=False, tol=0.0001, random_state=0) 
# Fit on training set
ridge.fit(X_train, y_train)
# Make predictions on validation set using best model from GridSearchCV
ridge_train_pred = ridge.predict(X_train)
ridge_val_pred = ridge.predict(X_validation)
# Calculate Forecast metrics on train set
mse_train_ridge = mean_squared_error(y_train, ridge_train_pred)
r2_train_ridge = r2_score(y_train, ridge_train_pred)
mae_train_ridge = mean_absolute_error(y_train, ridge_train_pred)
print("Train MSE: ", mse_train_ridge)
print("Train R2: ", r2_train_ridge)
print("Train MAE: ", mae_train_ridge)
# Calculate Forecast metrics on validation set
mse_val_ridge = mean_squared_error(y_validation, ridge_val_pred)
r2_val_ridge = r2_score(y_validation, ridge_val_pred)
mae_val_ridge = mean_absolute_error(y_validation, ridge_val_pred)
print("Validation MSE: ", mse_val_ridge)
print("Validation R2: ", r2_val_ridge)
print("Validation MAE: ", mae_val_ridge)

Train MSE:  0.018369171578226075
Train R2:  0.981630828421774
Train MAE:  0.10462058962635087
Validation MSE:  0.06279888654904399
Validation R2:  0.8010845286233097
Validation MAE:  0.19153127184422136


In [40]:
# Define parameter grid for GridSearchCV
#alphas_ridge = np.logspace(0.1, 5, num=500)
alphas_ridge = np.arange(1, 10, 0.05)
param_grid = {'alpha': alphas_ridge,
              #'positive': [True, False],
              'fit_intercept': [True, False],
              'random_state': [0]}

# Instantiate Ridge model
ridge = Ridge()
# Define GridSearchCV object
grid_search_ridge = GridSearchCV(ridge, param_grid, cv=tscv5, scoring='neg_mean_squared_error') 
# Fit GridSearchCV on training set
grid_search_ridge.fit(X_train, y_train)
# Print best parameter and score from GridSearchCV
print("Best parameter: ", grid_search_ridge.best_params_)
print("Best score: ", -grid_search_ridge.best_score_)
# Make predictions on validation set using best model from GridSearchCV
best_ridge = grid_search_ridge.best_estimator_
ridge_val_pred = best_ridge.predict(X_validation)
# Calculate Forecast metrics on validation set
mse_val_ridge = mean_squared_error(y_validation, ridge_val_pred)
r2_val_ridge = r2_score(y_validation, ridge_val_pred)
mae_val_ridge = mean_absolute_error(y_validation, ridge_val_pred)
print("Validation MSE: ", mse_val_ridge)
print("Validation R2: ", r2_val_ridge)
print("Validation MAE: ", mae_val_ridge)

Best parameter:  {'alpha': 9.950000000000008, 'fit_intercept': False, 'random_state': 0}
Best score:  0.11403563387203781
Validation MSE:  0.18375500698486313
Validation R2:  0.4179560205470315
Validation MAE:  0.35642540686993746


In [53]:
# Get the coefficients from the Ridge model
coef = ridge.coef_
# Create a dataframe of feature importances
feature_importance_ridge = pd.DataFrame({'Feature': X_train.columns, 'Importance': coef})
# Sort the features by importance
feature_importance_ridge = feature_importance_ridge.sort_values('Importance', ascending=False).reset_index(drop=True)
feature_importance_ridge.columns = ['feat_ridge', 'imp_ridge']
# Print the feature importances
feature_importance_ridge.head(15)

feat_ridge  imp_ridge
0        lag_1   0.389103
1     rice3_su   0.133199
2    squash_co   0.129054
3     flour_lp   0.104825
4       oil_cb   0.104084
5        lag_6   0.097503
6       silver   0.082430
7   chicken_tr   0.072616
8   platano_co   0.068520
9     oil2_bol   0.060541
10    sugar_lp   0.060107
11    sugar_tj   0.057815
12      oil_tr   0.056930
13       lag_2   0.051240
14  paprika_or   0.044100

### **3.2. Lasso**
Without tuning:
Validation MSE:  0.496
Validation R2:  -0.570
Validation MAE:  0.598

In [67]:
#lasso = Lasso(random_state=0)
lasso = Lasso(alpha=0.14, fit_intercept=True, max_iter=10000, tol=0.0001, positive=True, random_state=0)  
# Fit on training set
lasso.fit(X_train, y_train)
# Make predictions on validation set using best model from GridSearchCV
lasso_train_pred = lasso.predict(X_train)
lasso_val_pred = lasso.predict(X_validation)
# Calculate Forecast metrics on train set
mse_train_lasso = mean_squared_error(y_train, lasso_train_pred)
r2_train_lasso = r2_score(y_train, lasso_train_pred)
mae_train_lasso = mean_absolute_error(y_train, lasso_train_pred)
print("Train MSE: ", mse_train_lasso)
print("Train R2: ", r2_train_lasso)
print("Train MAE: ", mae_train_lasso)
# Calculate Forecast metrics on validation set
mse_val_lasso = mean_squared_error(y_validation, lasso_val_pred)
r2_val_lasso = r2_score(y_validation, lasso_val_pred)
mae_val_lasso = mean_absolute_error(y_validation, lasso_val_pred)
print("Validation MSE: ", mse_val_lasso)
print("Validation R2: ", r2_val_lasso)
print("Validation MAE: ", mae_val_lasso)

Train MSE:  0.07203978706743892
Train R2:  0.927960212932561
Train MAE:  0.20427867774468042
Validation MSE:  0.05601764619005393
Validation R2:  0.8225641072695601
Validation MAE:  0.17214351534568703


In [16]:
# Define parameter grid for GridSearchCV
alphas_lasso = np.logspace(-1, 2, num=500)
#alphas_lasso = np.arange(1, 100, 0.05)
param_grid = {'alpha': alphas_lasso,
              'positive': [True, False],
              'fit_intercept': [True, False],
              'random_state': [0]}

# Instantiate lasso model
lasso = Lasso()
# Define GridSearchCV object
grid_search_lasso = GridSearchCV(lasso, param_grid, cv=tscv5, scoring='neg_mean_squared_error')  
# Fit GridSearchCV on training set
grid_search_lasso.fit(X_train, y_train)
# Print best parameter and score from GridSearchCV
print("Best parameter: ", grid_search_lasso.best_params_)
print("Best score: ", -grid_search_lasso.best_score_)
# Make predictions on validation set using best model from GridSearchCV
best_lasso = grid_search_lasso.best_estimator_
y_val_pred = best_lasso.predict(X_validation)
# Calculate RMSE on validation set
mse_val = mean_squared_error(y_validation, y_val_pred)
r2_val = r2_score(y_validation, y_val_pred)
mae_val = mean_absolute_error(y_validation, y_val_pred)
print("Validation MSE: ", mse_val)
print("Validation R2: ", r2_val)
print("Validation MAE: ", mae_val)
# Make predictions on test set using best model from GridSearchCV
#y_test_pred_lasso = best_lasso.predict(X_test)

Best parameter:  {'alpha': 0.1, 'fit_intercept': False, 'positive': True, 'random_state': 0}
Best score:  0.05632981206453737
Validation MSE:  0.05000914094000738
Validation R2:  0.8415960474799797
Validation MAE:  0.15769631082399915


c:\Users\osmar.bolivar\AppData\Local\Programs\Python\Python310\lib\site-packages\numpy\ma\core.py:2820: RuntimeWarning: invalid value encountered in cast
  _data = np.array(data, dtype=dtype, copy=copy,


In [68]:
# Get the coefficients from the Lasso model
coef = lasso.coef_
# Create a dataframe of feature importances
feature_importance_lasso = pd.DataFrame({'Feature': X_train.columns, 'Importance': coef})
# Sort the features by importance
feature_importance_lasso = feature_importance_lasso.sort_values('Importance', ascending=False).reset_index(drop=True)
feature_importance_lasso.columns = ['feat_lasso', 'imp_lasso']
# Print the feature importances
feature_importance_lasso.head(15)

feat_lasso  imp_lasso
0          lag_1   0.791611
1       sugar_lp   0.043397
2         oil_tr   0.010839
3     chicken_po   0.000000
4   pineapple_tr   0.000000
5       rice2_lp   0.000000
6       rice3_su   0.000000
7            ufv   0.000000
8     paprika_or   0.000000
9      squash_co   0.000000
10      papa1_po   0.000000
11      rice2_po   0.000000
12      flour_lp   0.000000
13        silver   0.000000
14       milk_sc   0.000000

### **3.3. ElasticNet**
Without tuning:
Validation MSE:  0.2237
Validation R2:  0.250
Validation MAE:  0.401

In [84]:
#enet = ElasticNet()
enet = ElasticNet(alpha=0.9, l1_ratio=0.45, fit_intercept=False, positive=False) 
# Fit on training set
enet.fit(X_train, y_train)
# Make predictions on validation set using best model from GridSearchCV
enet_train_pred = enet.predict(X_train)
enet_val_pred = enet.predict(X_validation)
# Calculate Forecast metrics on train set
mse_train_enet = mean_squared_error(y_train, enet_train_pred)
r2_train_enet = r2_score(y_train, enet_train_pred)
mae_train_enet = mean_absolute_error(y_train, enet_train_pred)
print("Train MSE: ", mse_train_enet)
print("Train R2: ", r2_train_enet)
print("Train MAE: ", mae_train_enet)
# Calculate Forecast metrics on validation set
mse_val_enet = mean_squared_error(y_validation, enet_val_pred)
r2_val_enet = r2_score(y_validation, enet_val_pred)
mae_val_enet = mean_absolute_error(y_validation, enet_val_pred)
print("Validation MSE: ", mse_val_enet)
print("Validation R2: ", r2_val_enet)
print("Validation MAE: ", mae_val_enet)

Train MSE:  0.3346501972327239
Train R2:  0.6653498027672762
Train MAE:  0.45091808751332985
Validation MSE:  0.20699349562391822
Validation R2:  0.3443481084368465
Validation MAE:  0.38118731133367745


In [ ]:
# Define parameter grid for GridSearchCV
alphas_enet = np.logspace(-1, 2, num=500)
#alphas_enet = np.arange(1, 100, 0.05)
param_grid = {'alpha': alphas_enet,
              'l1_ratio': np.arange(0.01, 0.95, 0.01),
              'positive': [True, False],
              'fit_intercept': [True, False],
              'random_state': [0]}

# Instantiate enet model
enet = ElasticNet()
# Define GridSearchCV object
grid_search_enet = GridSearchCV(enet, param_grid, cv=tscv5, scoring='neg_mean_squared_error') 
# Fit GridSearchCV on training set
grid_search_enet.fit(X_train, y_train)
# Print best parameter and score from GridSearchCV
print("Best parameter: ", grid_search_enet.best_params_)
print("Best score: ", -grid_search_enet.best_score_)
# Make predictions on validation set using best model from GridSearchCV
best_enet = grid_search_enet.best_estimator_
y_val_pred = best_enet.predict(X_validation)
# Calculate RMSE on validation set
mse_val = mean_squared_error(y_validation, y_val_pred)
r2_val = r2_score(y_validation, y_val_pred)
mae_val = mean_absolute_error(y_validation, y_val_pred)
print("Validation MSE: ", mse_val)
print("Validation R2: ", r2_val)
print("Validation MAE: ", mae_val)
# Make predictions on test set using best model from GridSearchCV
#y_test_pred_enet = best_enet.predict(X_test)

In [33]:
# Get the coefficients from the enet model
coef = enet.coef_
# Create a dataframe of feature importances
feature_importance_enet = pd.DataFrame({'Feature': X_train.columns, 'Importance': coef})
# Sort the features by importance
feature_importance_enet = feature_importance_enet.sort_values('Importance', ascending=False).reset_index(drop=True)
feature_importance_enet.columns = ['feat_enet', 'imp_enet']
# Print the feature importances
feature_importance_enet.head(15)

feat_enet  imp_enet
0      lag_1  0.223097
1      lag_2  0.117231
2      lag_3  0.016951
3      PC_58 -0.000000
4      PC_56 -0.000000
5      PC_55  0.000000
6      PC_54 -0.000000
7      PC_53  0.000000
8      PC_52  0.000000
9      PC_51  0.000000
10     PC_50  0.000000
11     PC_49 -0.000000
12     PC_48 -0.000000
13     PC_47  0.000000
14     PC_46 -0.000000

### **3.4. ADA**
Without tuning:
Validation MSE:  0.106
Validation R2:  0.663
Validation MAE:  0.247

In [95]:
#ada = AdaBoostRegressor(random_state=0)
ada = AdaBoostRegressor(estimator=DecisionTreeRegressor(max_depth=6), random_state=0)
# Fit the model on training data
ada.fit(X_train, y_train)
# Make predictions on the validation set
ada_train_pred = ada.predict(X_train)
ada_val_pred = ada.predict(X_validation)
# Calculate Forecast metrics on train set
mse_train_ada = mean_squared_error(y_train, ada_train_pred)
r2_train_ada = r2_score(y_train, ada_train_pred)
mae_train_ada = mean_absolute_error(y_train, ada_train_pred)
print("Train MSE: ", mse_train_ada)
print("Train R2: ", r2_train_ada)
print("Train MAE: ", mae_train_ada)
# Evaluate the model on the validation set
mse_val_ada = mean_squared_error(y_validation, ada_val_pred)
r2_val_ada = r2_score(y_validation, ada_val_pred)
mae_val_ada = mean_absolute_error(y_validation, ada_val_pred)
print("Validation MSE: ", mse_val_ada)
print("Validation R2: ", r2_val_ada)
print("Validation MAE: ", mae_val_ada)

Train MSE:  0.0013178660371355355
Train R2:  0.9986821339628644
Train MAE:  0.02358666665543112
Validation MSE:  0.05433635731461519
Validation R2:  0.8278895897351943
Validation MAE:  0.2049995735352874


In [90]:
# Define the AdaBoost Regressor
ada = AdaBoostRegressor()
# Define the range of hyperparameters to search over
param_grid_ada = {
    #'n_estimators': range(50, 200, 5),    ##120 was selected range(50, 200, 5)
    #'learning_rate': np.logspace(-5,1, 100),
    #'loss': ['linear', 'square', 'exponential']
    'random_state': [0],
    'estimator': [DecisionTreeRegressor(max_depth=3),
                  DecisionTreeRegressor(max_depth=4),
                  DecisionTreeRegressor(max_depth=5),
                  DecisionTreeRegressor(max_depth=6),
                  DecisionTreeRegressor(max_depth=7),
                  DecisionTreeRegressor(max_depth=8),
                  ]
}

# Tune hyperparameters using GridSearchCV with TimeSeriesSplit
grid_search_ada = GridSearchCV(estimator=ada, param_grid=param_grid_ada, cv=tscv5, scoring='neg_mean_squared_error')
grid_search_ada.fit(X_train, y_train)
# Evaluate the model using the best hyperparameters on the test set
ada_best = AdaBoostRegressor(**grid_search_ada.best_params_)
ada_best.fit(X_train, y_train)
# Make predictions on the validation set
y_val_pred = ada_best.predict(X_validation)
# Evaluate the model on the validation set
mse_val = mean_squared_error(y_validation, y_val_pred)
r2_val = r2_score(y_validation, y_val_pred)
mae_val = mean_absolute_error(y_validation, y_val_pred)
print("Validation MSE: ", mse_val)
print("Validation R2: ", r2_val)
print("Validation MAE: ", mae_val)
# Print the best hyperparameters and the best score
print("Best parameters found: ", grid_search_ada.best_params_)
print("Lowest MSE found: ", -grid_search_ada.best_score_)
# Evaluate the model on the test set
#y_test_pred_ada = ada_best.predict(X_test)

Validation MSE:  0.11483229244788656
Validation R2:  0.6362687169031497
Validation MAE:  0.2893540072845322
Best parameters found:  {'estimator': DecisionTreeRegressor(max_depth=5), 'random_state': 0}
Lowest MSE found:  0.1547741727368356


In [96]:
# Create a DataFrame with the feature importance values
feature_importance_ada = pd.DataFrame({'Feature': X_train.columns, 'Importance': ada.feature_importances_})
# Sort the DataFrame by importance values in descending order
feature_importance_ada = feature_importance_ada.sort_values(by='Importance', ascending=False).reset_index(drop=True)
feature_importance_ada.columns = ['feat', 'imp_ada']
# Print the feature importance DataFrame
feature_importance_ada.head(15)

feat   imp_ada
0         lag_1  0.600317
1      oil2_bol  0.100222
2    chicken_tj  0.073183
3      sugar_tj  0.050236
4      sugar_lp  0.029374
5   chicken_bol  0.025084
6    paprika_or  0.017356
7      rice2_po  0.008339
8      flour_lp  0.007969
9      sugar_cb  0.007457
10     papa1_po  0.006715
11          ufv  0.006302
12     ycorn_sc  0.005591
13     ycorn_tj  0.005347
14        libor  0.005242

### **3.5. GBR**
Without tuning:
Validation MSE:  0.080
Validation R2:  0.746
Validation MAE:  0.203

In [116]:
#gbr = GradientBoostingRegressor(random_state=0)
gbr = GradientBoostingRegressor(learning_rate=0.04, n_estimators=275, random_state=0)
# Fit the model on the training set
gbr.fit(X_train, y_train)
# Make predictions on validation set using best model from GridSearchCV
gbr_train_pred = gbr.predict(X_train)
gbr_val_pred = gbr.predict(X_validation)
# Calculate Forecast metrics on train set
mse_train_gbr = mean_squared_error(y_train, gbr_train_pred)
r2_train_gbr = r2_score(y_train, gbr_train_pred)
mae_train_gbr = mean_absolute_error(y_train, gbr_train_pred)
print("Train MSE: ", mse_train_gbr)
print("Train R2: ", r2_train_gbr)
print("Train MAE: ", mae_train_gbr)
# Calculate Forecast metrics on validation set
mse_val_gbr = mean_squared_error(y_validation, gbr_val_pred)
r2_val_gbr = r2_score(y_validation, gbr_val_pred)
mae_val_gbr = mean_absolute_error(y_validation, gbr_val_pred)
print("Validation MSE: ", mse_val_gbr)
print("Validation R2: ", r2_val_gbr)
print("Validation MAE: ", mae_val_gbr)

Train MSE:  0.0005383824581009735
Train R2:  0.999461617541899
Train MAE:  0.01850733503960401
Validation MSE:  0.053465345503517954
Validation R2:  0.8306485196223311
Validation MAE:  0.1982033171340982


In [99]:
# Define the model
gbr = GradientBoostingRegressor()

# Define the hyperparameters to be tuned
params = {
    'learning_rate': np.logspace(-2,0.5, 50),
    #'loss': ['squared_error', 'absolute_error', 'huber', 'quantile'],   ## 'loss': 'squared_error'
    'n_estimators': range(100, 300, 10),
    #'max_depth': range(3,6,1),
    #'min_samples_split': range(2,20,1),
    #'min_samples_leaf': range(1,50, 1),
    #'min_weight_fraction_leaf': np.arange(0.0, 0.5, 0.01),
    #'subsample': np.arange(0.5, 1, 0.05),
    #'max_features': [None, 'sqrt', 'log2'],
    #'max_leaf_nodes': range(2, 200, 1),
    #'criterion': ['friedman_mse', 'squared_error'],
    'random_state': [0]
}

# Create the GridSearchCV object
grid_gbr = GridSearchCV(gbr, params, cv=tscv5, scoring='neg_mean_squared_error', n_jobs=-1)
# Fit the model on the training set with GridSearchCV
grid_gbr.fit(X_train, y_train)
# Print the best hyperparameters
print('Best hyperparameters:', grid_gbr.best_params_)
# Use the best model to make predictions on the validation set
y_val_pred = grid_gbr.predict(X_validation)
# Compute the mean squared error of the predictions on the validation set
mse_val = mean_squared_error(y_validation, y_val_pred)
r2_val = r2_score(y_validation, y_val_pred)
mae_val = mean_absolute_error(y_validation, y_val_pred)
print("Validation MSE: ", mse_val)
print("Validation R2: ", r2_val)
print("Validation MAE: ", mae_val)
# Use the best model to make predictions on the test set
#y_test_pred_gbr = grid_gbr.predict(X_test)

c:\Users\osmar\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\model_selection\_search.py:1119: RuntimeWarning: overflow encountered in square
  (array - array_means[:, np.newaxis]) ** 2, axis=1, weights=weights


Best hyperparameters: {'learning_rate': 0.6866488450043002, 'n_estimators': 110, 'random_state': 0}
Validation MSE:  0.0911476260721347
Validation R2:  0.7112898969817645
Validation MAE:  0.23047731139223362


In [117]:
# Create a DataFrame with the feature importance values
feature_importance_gbr = pd.DataFrame({'Feature': X_train.columns, 'Importance': gbr.feature_importances_})
# Sort the DataFrame by importance values in descending order
feature_importance_gbr = feature_importance_gbr.sort_values(by='Importance', ascending=False).reset_index(drop=True)
feature_importance_gbr.columns = ['feat_gbr', 'imp_gbr']
# Print the feature importance DataFrame
feature_importance_gbr.head(15)

feat_gbr   imp_gbr
0        lag_1  0.727985
1     oil2_bol  0.090595
2   chicken_tj  0.061281
3     sugar_tj  0.029617
4     sugar_lp  0.014371
5     papa1_po  0.007847
6     ycorn_sc  0.006817
7        libor  0.005696
8     ycorn_tj  0.005398
9    tomato_tj  0.004835
10    sugar_po  0.004362
11         ufv  0.004340
12    flour_lp  0.004249
13    rice2_po  0.003931
14      lag_12  0.003802

### **3.4. RF**
Without tuning:
Validation MSE:  0.085
Validation R2:  0.730
Validation MAE:  0.247

In [121]:
# Define the Random Forest Regression model
rf = RandomForestRegressor(random_state=0)
#rf = RandomForestRegressor(n_estimators=265, random_state=0)

# Fit the model to the training data and make predictions on the validation set
rf.fit(X_train, y_train)
# Make predictions on validation set using best model from GridSearchCV
rf_train_pred = rf.predict(X_train)
rf_val_pred = rf.predict(X_validation)
# Calculate Forecast metrics on train set
mse_train_rf = mean_squared_error(y_train, rf_train_pred)
r2_train_rf = r2_score(y_train, rf_train_pred)
mae_train_rf = mean_absolute_error(y_train, rf_train_pred)
print("Train MSE: ", mse_train_rf)
print("Train R2: ", r2_train_rf)
print("Train MAE: ", mae_train_rf)
# Calculate Forecast metrics on validation set
mse_val_rf = mean_squared_error(y_validation, rf_val_pred)
r2_val_rf = r2_score(y_validation, rf_val_pred)
mae_val_rf = mean_absolute_error(y_validation, rf_val_pred)
print("Validation MSE: ", mse_val_rf)
print("Validation R2: ", r2_val_rf)
print("Validation MAE: ", mae_val_rf)

Train MSE:  0.006231533078617847
Train R2:  0.9937684669213821
Train MAE:  0.059591666937122624
Validation MSE:  0.05964256772046444
Validation R2:  0.8110821684240043
Validation MAE:  0.204859244375391


In [ ]:
# Define the Random Forest Regression model
rf_reg = RandomForestRegressor()

# Define the hyperparameters to tune
param_grid_rf = {
    'n_estimators': range(100, 300, 15),
    #'max_features': [None, 'sqrt', 'sqrt']
    #'max_depth': range(3,7,1),
    'min_samples_split': range(2,20,1),
    #'min_samples_leaf': range(1,50, 1),
    #'min_weight_fraction_leaf': np.arange(0.0, 0.5, 0.01),
    #'bootstrap': [True],
    #'oob_score': [True, False],
    #'warm_start': [True, False],
    #'max_samples': np.arange(0.1, 1.0, 0.01)
    'random_state': [0]
}

# Define the GridSearchCV object
grid_rf_reg = GridSearchCV(estimator=rf_reg, param_grid=param_grid_rf, cv=tscv5, scoring='neg_mean_squared_error')
# Fit the GridSearchCV object to the training data
grid_rf_reg.fit(X_train, y_train)
# Extract the best hyperparameters and score
best_params = grid_rf_reg.best_params_
best_score = grid_rf_reg.best_score_
# Print the best hyperparameters found by GridSearchCV
print(f"Best hyperparameters: {best_params}")
print(f"Best score: {best_score}")
# Instantiate a new Random Forest Regression model using the best hyperparameters
rf_reg_best = RandomForestRegressor(**best_params)
# Fit the model to the training data and make predictions on the validation set
rf_reg_best.fit(X_train, y_train)
# Use the best model to make predictions on the validation set
y_val_pred = rf_reg_best.predict(X_validation)
# Compute the mean squared error of the predictions on the validation set
mse_val = mean_squared_error(y_validation, y_val_pred)
r2_val = r2_score(y_validation, y_val_pred)
mae_val = mean_absolute_error(y_validation, y_val_pred)
print("Validation MSE: ", mse_val)
print("Validation R2: ", r2_val)
print("Validation MAE: ", mae_val)

# Use the best model to make predictions on the test set
#y_test_pred_rf = rf_reg_best.predict(X_test)

Best hyperparameters: {'min_samples_split': 2, 'n_estimators': 265, 'random_state': 0}
Best score: -0.0035817383690906874
Validation MSE:  0.0009983900502684609
Validation R2:  0.9989484444515393
Validation MAE:  0.024970310604635504


In [39]:
# Create a DataFrame with the feature importance values
feature_importance_rf = pd.DataFrame({'Feature': X_train.columns, 'Importance': rf.feature_importances_})
# Sort the DataFrame by importance values in descending order
feature_importance_rf = feature_importance_rf.sort_values(by='Importance', ascending=False).reset_index(drop=True)
feature_importance_rf.columns = ['feat', 'imp_rf']
# Print the feature importance DataFrame
feature_importance_rf.head(30)

feat    imp_rf
0    lag_1  0.876660
1     PC_2  0.025637
2     PC_3  0.023071
3    PC_11  0.006162
4   lag_12  0.005769
5     PC_6  0.003004
6    PC_10  0.002788
7     PC_4  0.002594
8     PC_9  0.002449
9    lag_6  0.001935
10   PC_44  0.001927
11   PC_66  0.001911
12    PC_5  0.001801
13   lag_2  0.001732
14   PC_49  0.001576
15   PC_31  0.001499
16   PC_77  0.001427
17   PC_74  0.001203
18   PC_39  0.001161
19   PC_21  0.001147
20   PC_32  0.001133
21   PC_41  0.001041
22    PC_1  0.001033
23   PC_37  0.001003
24   PC_18  0.000965
25   PC_60  0.000930
26   PC_79  0.000863
27   PC_33  0.000842
28   PC_71  0.000774
29   PC_43  0.000754

### **3.5. ET**   
Without tuning:
Validation MSE:  0.212
Validation R2:  0.330
Validation MAE:  0.396

In [23]:
# Define the Extra Trees Regression model
et = ExtraTreesRegressor(random_state=0)
#et = ExtraTreesRegressor(bootstrap=True, max_samples=0.9599999999999995, oob_score=True, random_state=0)
# Fit the model to the training data and make predictions on the validation set
et.fit(X_train, y_train)
# Make predictions on validation set using best model from GridSearchCV
et_train_pred = et.predict(X_train)
et_val_pred = et.predict(X_validation)
# Calculate Forecast metrics on train set
mse_train_et = mean_squared_error(y_train, et_train_pred)
r2_train_et = r2_score(y_train, et_train_pred)
mae_train_et = mean_absolute_error(y_train, et_train_pred)
print("Train MSE: ", mse_train_et)
print("Train R2: ", r2_train_et)
print("Train MAE: ", mae_train_et)
# Calculate Forecast metrics on validation set
mse_val_et = mean_squared_error(y_validation, et_val_pred)
r2_val_et = r2_score(y_validation, et_val_pred)
mae_val_et = mean_absolute_error(y_validation, et_val_pred)
print("Validation MSE: ", mse_val_et)
print("Validation R2: ", r2_val_et)
print("Validation MAE: ", mae_val_et)

Train MSE:  1.6732225052706325e-30
Train R2:  1.0
Train MAE:  8.704402613401371e-16
Validation MSE:  0.08720435579558937
Validation R2:  0.7237802054717427
Validation MAE:  0.21167218133499283


In [ ]:
# Define the Extra Trees Regression model
et_reg = ExtraTreesRegressor()

# Define the hyperparameter grid to search over
param_grid = {
    #'n_estimators': range(100, 300, 15),
    #'max_depth': range(3,15,1),
    #'max_features': [None, 'sqrt', 'sqrt']
    'min_samples_split': range(2,20,1),
    'min_samples_leaf': range(1,50, 1),
    #'min_weight_fraction_leaf': np.arange(0.0, 0.5, 0.01),
    'bootstrap': [True],
    'oob_score': [True, False],
    #'warm_start': [True, False],
    'max_samples': np.arange(0.1, 1.0, 0.01),
    #'criterion': ['squared_error', 'absolute_error', 'friedman_mse'],
    #'min_impurity_decrease': np.arange(0.0, 0.01, 0.00001),
    'random_state': [0]
}

# Use GridSearchCV to find the best hyperparameters
grid_search = GridSearchCV(et_reg, param_grid=param_grid, cv=tscv5, scoring='neg_mean_squared_error')
grid_search.fit(X_train, y_train)
# Print the best hyperparameters and the corresponding validation score
print("Best hyperparameters: ", grid_search.best_params_)
#print("Validation score: ", grid_search.best_score_)
# Use the best model to make predictions on the validation set
best_et_reg = grid_search.best_estimator_
y_val_pred = best_et_reg.predict(X_validation)
# Compute the mean squared error of the predictions on the validation set
mse_val = mean_squared_error(y_validation, y_val_pred)
r2_val = r2_score(y_validation, y_val_pred)
mae_val = mean_absolute_error(y_validation, y_val_pred)
print("Validation MSE: ", mse_val)
print("Validation R2: ", r2_val)
print("Validation MAE: ", mae_val)

# Use the best model to make predictions on the test set
#y_test_pred_et = best_et_reg.predict(X_test)

In [41]:
# Create a DataFrame with the feature importance values
feature_importance_et = pd.DataFrame({'Feature': X_train.columns, 'Importance': et.feature_importances_})
# Sort the DataFrame by importance values in descending order
feature_importance_et = feature_importance_et.sort_values(by='Importance', ascending=False).reset_index(drop=True)
feature_importance_et.columns = ['feat', 'imp_et']
# Print the feature importance DataFrame
feature_importance_et.head(30)

feat    imp_et
0    lag_1  0.467694
1    lag_2  0.195290
2     PC_2  0.136688
3    lag_3  0.039389
4    PC_11  0.014717
5     PC_3  0.014463
6   lag_12  0.013275
7    lag_6  0.012449
8    lag_9  0.006027
9     PC_5  0.005856
10    PC_1  0.005045
11    PC_6  0.004609
12   PC_10  0.004241
13    PC_9  0.004106
14   PC_25  0.002872
15    PC_4  0.002740
16   PC_21  0.002478
17    PC_8  0.002234
18   PC_15  0.002133
19   PC_29  0.001839
20   PC_62  0.001794
21   PC_24  0.001777
22   PC_44  0.001741
23   PC_58  0.001672
24   PC_31  0.001654
25   PC_22  0.001632
26   PC_17  0.001624
27   PC_37  0.001588
28   PC_16  0.001468
29   PC_23  0.001457

## **4. Report**

#### End